File: mmf/framework/push/webhook.py
"""
Webhook Push Adapter.

HTTP webhook adapter for delivering notifications to external endpoints.
Supports HMAC signing, circuit breaker pattern, and configurable retries.

Features:
- HMAC-SHA256 payload signing
- Circuit breaker pattern for endpoint protection
- Exponential backoff retry
- Event type filtering
- Configurable timeouts
"""

from __future__ import annotations

import asyncio
import hashlib
import hmac
import json
import logging
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from mmf.core.push import IPushAdapter, PushChannel, PushMessage, PushResult, PushStatus

logger = logging.getLogger(__name__)


# =============================================================================
# Configuration
# =============================================================================


@dataclass
class WebhookEndpointConfig:
    """Configuration for a single webhook endpoint."""

    url: str
    secret: str = ""  # Empty = no signature
    event_types: list[str] = field(default_factory=list)  # Empty = all events
    enabled: bool = True
    custom_headers: dict[str, str] = field(default_factory=dict)


@dataclass
class WebhookConfig:
    """Webhook adapter configuration."""

    # Timeout settings
    connect_timeout: float = 5.0
    read_timeout: float = 30.0

    # Retry settings
    max_retries: int = 3
    initial_backoff: float = 1.0
    max_backoff: float = 60.0

    # Circuit breaker settings
    failure_threshold: int = 5
    recovery_timeout: int = 300  # 5 minutes

    # Signature settings
    signature_header: str = "X-MMF-Signature"
    event_header: str = "X-MMF-Event"
    delivery_id_header: str = "X-MMF-Delivery-Id"
    timestamp_header: str = "X-MMF-Timestamp"


# =============================================================================
# Circuit Breaker
# =============================================================================


@dataclass
class CircuitBreakerState:
    """Circuit breaker state for an endpoint."""

    failures: int = 0
    last_failure: datetime | None = None
    is_open: bool = False

    def record_failure(self, threshold: int) -> None:
        """Record a failure and potentially open the circuit."""
        self.failures += 1
        self.last_failure = datetime.now(timezone.utc)
        if self.failures >= threshold:
            self.is_open = True
            logger.warning(f"Circuit breaker opened after {self.failures} failures")

    def record_success(self) -> None:
        """Record a success and reset the circuit."""
        if self.is_open:
            logger.info("Circuit breaker closed after successful request")
        self.failures = 0
        self.is_open = False

    def should_allow(self, recovery_timeout: int) -> bool:
        """Check if a request should be allowed."""
        if not self.is_open:
            return True

        # Check if recovery timeout has elapsed
        if self.last_failure:
            elapsed = (datetime.now(timezone.utc) - self.last_failure).total_seconds()
            if elapsed >= recovery_timeout:
                return True  # Allow one attempt (half-open)

        return False


# =============================================================================
# Webhook Adapter
# =============================================================================


class WebhookAdapter:
    """
    Webhook push adapter.

    Implements IPushAdapter for HTTP webhook delivery. Sends push
    messages as JSON payloads to configured endpoints.

    Usage:
        config = WebhookConfig(max_retries=5)
        adapter = WebhookAdapter(config)
        await adapter.start()

        # Send with endpoints in target
        message = PushMessage(
            target=PushTarget(webhook_urls=["https://example.com/webhook"]),
            data={"event": "notification"},
        )
        result = await adapter.send(message)

        # Or with explicit endpoint configs
        result = await adapter.send_to_endpoints(
            message,
            endpoints=[
                WebhookEndpointConfig(
                    url="https://example.com/webhook",
                    secret="my-secret",  # pragma: allowlist secret
                )
            ],
        )
    """

    def __init__(self, config: WebhookConfig | None = None):
        """
        Initialize the webhook adapter.

        Args:
            config: Webhook configuration (uses defaults if not provided)
        """
        self.config = config or WebhookConfig()
        self._client: Any = None  # httpx.AsyncClient
        self._circuit_breakers: dict[str, CircuitBreakerState] = {}

    @property
    def channel(self) -> PushChannel:
        """The channel this adapter handles."""
        return PushChannel.WEBHOOK

    async def start(self) -> None:
        """Start the webhook adapter."""
        try:
            import httpx

            self._client = httpx.AsyncClient(
                timeout=httpx.Timeout(
                    connect=self.config.connect_timeout,
                    read=self.config.read_timeout,
                    write=self.config.read_timeout,
                    pool=self.config.connect_timeout,
                ),
            )
            logger.info("Webhook adapter started")
        except ImportError:
            raise RuntimeError("httpx is required for webhook adapter: pip install httpx")

    async def stop(self) -> None:
        """Stop the webhook adapter and cleanup resources."""
        if self._client is not None:
            await self._client.aclose()
            self._client = None
        logger.info("Webhook adapter stopped")

    def _get_circuit_breaker(self, url: str) -> CircuitBreakerState:
        """Get or create circuit breaker for an endpoint."""
        if url not in self._circuit_breakers:
            self._circuit_breakers[url] = CircuitBreakerState()
        return self._circuit_breakers[url]

    async def send(self, message: PushMessage) -> PushResult:
        """
        Send a push message to webhook endpoints.

        Uses URLs from message.target.webhook_urls.

        Args:
            message: The push message

        Returns:
            PushResult with delivery status
        """
        if not message.target.webhook_urls:
            return PushResult(
                message_id=message.id,
                channel=PushChannel.WEBHOOK,
                status=PushStatus.FAILED,
                success=False,
                error_code="NO_ENDPOINTS",
                error_message="No webhook URLs provided",
            )

        # Convert URLs to endpoint configs
        endpoints = [WebhookEndpointConfig(url=url) for url in message.target.webhook_urls]

        return await self.send_to_endpoints(message, endpoints)

    async def send_to_endpoints(
        self,
        message: PushMessage,
        endpoints: list[WebhookEndpointConfig],
    ) -> PushResult:
        """
        Send a push message to specific webhook endpoints.

        Args:
            message: The push message
            endpoints: List of endpoint configurations

        Returns:
            PushResult with delivery status
        """
        if not endpoints:
            return PushResult(
                message_id=message.id,
                channel=PushChannel.WEBHOOK,
                status=PushStatus.FAILED,
                success=False,
                error_code="NO_ENDPOINTS",
                error_message="No webhook endpoints provided",
            )

        # Filter by enabled and event type
        event_type = message.data.get("event_type", "")
        filtered = [
            ep
            for ep in endpoints
            if ep.enabled and (not ep.event_types or event_type in ep.event_types)
        ]

        if not filtered:
            return PushResult(
                message_id=message.id,
                channel=PushChannel.WEBHOOK,
                status=PushStatus.DELIVERED,
                success=True,
                metadata={"skipped": "No matching endpoints for event type"},
            )

        # Send to all endpoints in parallel
        tasks = [self._deliver_to_endpoint(message, ep) for ep in filtered]
        results = await asyncio.gather(*tasks, return_exceptions=True)

        # Aggregate results
        success_count = sum(1 for r in results if isinstance(r, PushResult) and r.success)
        total = len(filtered)

        return PushResult(
            message_id=message.id,
            channel=PushChannel.WEBHOOK,
            status=PushStatus.DELIVERED if success_count > 0 else PushStatus.FAILED,
            success=success_count > 0,
            delivered_at=datetime.now(timezone.utc) if success_count > 0 else None,
            metadata={
                "total_endpoints": total,
                "success_count": success_count,
                "failure_count": total - success_count,
            },
        )

    async def send_batch(self, messages: list[PushMessage]) -> list[PushResult]:
        """Send multiple messages."""
        results = []
        for message in messages:
            result = await self.send(message)
            results.append(result)
        return results

    async def _deliver_to_endpoint(
        self,
        message: PushMessage,
        endpoint: WebhookEndpointConfig,
    ) -> PushResult:
        """Deliver to a single endpoint with retries."""
        # Check circuit breaker
        circuit = self._get_circuit_breaker(endpoint.url)
        if not circuit.should_allow(self.config.recovery_timeout):
            return PushResult(
                message_id=message.id,
                channel=PushChannel.WEBHOOK,
                status=PushStatus.FAILED,
                success=False,
                error_code="CIRCUIT_OPEN",
                error_message="Circuit breaker is open for this endpoint",
            )

        # Build request body
        body = json.dumps(message.to_dict()).encode()

        # Generate signature
        signature = self._sign_payload(body, endpoint.secret)

        # Build headers
        headers = {
            "Content-Type": "application/json",
            self.config.event_header: message.data.get("event_type", "push"),
            self.config.delivery_id_header: message.id,
            self.config.timestamp_header: datetime.now(timezone.utc).isoformat(),
            **endpoint.custom_headers,
        }

        if signature:
            headers[self.config.signature_header] = signature

        if message.correlation_id:
            headers["X-MMF-Correlation-Id"] = message.correlation_id

        # Attempt delivery with retries
        for attempt in range(self.config.max_retries):
            try:
                if self._client is None:
                    await self.start()

                response = await self._client.post(
                    endpoint.url,
                    content=body,
                    headers=headers,
                )

                if response.status_code < 400:
                    circuit.record_success()
                    return PushResult(
                        message_id=message.id,
                        channel=PushChannel.WEBHOOK,
                        status=PushStatus.DELIVERED,
                        success=True,
                        delivered_at=datetime.now(timezone.utc),
                        attempt_number=attempt + 1,
                        metadata={
                            "status_code": response.status_code,
                            "endpoint": endpoint.url,
                        },
                    )

                # Server error - retry
                if response.status_code >= 500:
                    circuit.record_failure(self.config.failure_threshold)
                    backoff = min(
                        self.config.initial_backoff * (2**attempt),
                        self.config.max_backoff,
                    )
                    logger.warning(
                        f"Webhook {endpoint.url} returned {response.status_code}, "
                        f"retry {attempt + 1}, backing off {backoff}s"
                    )
                    await asyncio.sleep(backoff)
                    continue

                # Client error - don't retry
                return PushResult(
                    message_id=message.id,
                    channel=PushChannel.WEBHOOK,
                    status=PushStatus.REJECTED,
                    success=False,
                    error_code=str(response.status_code),
                    error_message=response.text[:200] if response.text else None,
                    attempt_number=attempt + 1,
                )

            except Exception as e:
                circuit.record_failure(self.config.failure_threshold)
                logger.error(f"Webhook error for {endpoint.url}: {e}")

                if attempt < self.config.max_retries - 1:
                    await asyncio.sleep(self.config.initial_backoff * (2**attempt))
                    continue

                return PushResult(
                    message_id=message.id,
                    channel=PushChannel.WEBHOOK,
                    status=PushStatus.FAILED,
                    success=False,
                    error_code="EXCEPTION",
                    error_message=str(e),
                    should_retry=True,
                )

        return PushResult(
            message_id=message.id,
            channel=PushChannel.WEBHOOK,
            status=PushStatus.FAILED,
            success=False,
            error_code="MAX_RETRIES",
            error_message="Max retries exceeded",
        )

    def _sign_payload(self, body: bytes, secret: str) -> str:
        """Generate HMAC-SHA256 signature for payload."""
        if not secret:
            return ""

        signature = hmac.new(
            secret.encode(),
            body,
            hashlib.sha256,
        ).hexdigest()

        return f"sha256={signature}"

    @staticmethod
    def verify_signature(body: bytes, secret: str, signature: str) -> bool:
        """
        Verify webhook signature.

        Args:
            body: Raw request body bytes
            secret: Webhook secret
            signature: Signature header value

        Returns:
            True if signature is valid
        """
        if not signature.startswith("sha256="):
            return False

        expected = f"sha256={hmac.new(secret.encode(), body, hashlib.sha256).hexdigest()}"
        return hmac.compare_digest(expected, signature)

    def get_circuit_breaker_stats(self) -> dict[str, Any]:
        """Get circuit breaker statistics for all endpoints."""
        return {
            url: {
                "failures": cb.failures,
                "is_open": cb.is_open,
                "last_failure": cb.last_failure.isoformat() if cb.last_failure else None,
            }
            for url, cb in self._circuit_breakers.items()
        }

File: crates/mmf-push/src/webhook.rs
use std::collections::BTreeMap;
use std::sync::{Arc, Mutex, RwLock};
use std::time::Instant;

use async_trait::async_trait;
use hmac::{Hmac, Mac};
use mmf_resilience::{
    BackoffPolicy, CircuitBreaker, CircuitBreakerConfig, CircuitState, ConfiguredBackoff,
    RetryConfig, RetryStrategy,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sha2::Sha256;

use crate::{
    PushAdapter, PushAdapterHealth, PushChannel, PushError, PushHealthStatus, PushMessage,
    PushResult, PushStatus, lifecycle::encode_hex,
};

type HmacSha256 = Hmac<Sha256>;
const TOKEN_MARKER: &str = "__MARTY_TOKEN__";

/// Tenant-scoped allowlist for callback destinations.
///
/// Registrations use `tenant|URL` entries separated by semicolons. A URL may
/// contain one `__MARTY_TOKEN__` slot matching 16-512 URL-safe characters.
#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct WebhookDestinationRegistry {
    registrations: BTreeMap<String, Vec<String>>,
}

impl WebhookDestinationRegistry {
    pub fn parse(configuration: &str) -> Result<Self, PushError> {
        let mut registrations = BTreeMap::<String, Vec<String>>::new();
        for raw_entry in configuration.split(';') {
            let entry = raw_entry.trim();
            if entry.is_empty() {
                continue;
            }
            let (tenant, template) = entry.split_once('|').ok_or_else(|| {
                PushError::InvalidConfiguration(
                    "webhook destinations must use tenant|URL entries".into(),
                )
            })?;
            let tenant = tenant.trim();
            let template = template.trim();
            if tenant.is_empty() || template.is_empty() {
                return Err(PushError::InvalidConfiguration(
                    "webhook destination tenant and URL must not be empty".into(),
                ));
            }
            if template.match_indices(TOKEN_MARKER).count() > 1 {
                return Err(PushError::InvalidConfiguration(
                    "webhook destination allows at most one token slot".into(),
                ));
            }
            validate_destination_shape(template)?;
            registrations
                .entry(tenant.to_owned())
                .or_default()
                .push(template.to_owned());
        }
        Ok(Self { registrations })
    }

    pub fn require(&self, tenant: &str, destination: &str) -> Result<(), PushError> {
        validate_destination_shape(destination)?;
        if self.registrations.get(tenant).is_some_and(|templates| {
            templates
                .iter()
                .any(|template| destination_matches(destination, template))
        }) {
            Ok(())
        } else {
            Err(PushError::InvalidOperation(
                "webhook destination is not registered for the tenant".into(),
            ))
        }
    }

    #[must_use]
    pub fn templates(&self, tenant: &str) -> &[String] {
        self.registrations.get(tenant).map_or(&[], Vec::as_slice)
    }
}

#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
pub struct WebhookEndpointConfig {
    pub url: String,
    #[serde(default, skip_serializing)]
    pub secret: String,
    #[serde(default)]
    pub event_types: Vec<String>,
    #[serde(default = "default_true")]
    pub enabled: bool,
    #[serde(default)]
    pub custom_headers: BTreeMap<String, String>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(default)]
pub struct WebhookConfig {
    pub connect_timeout_ms: u64,
    pub read_timeout_ms: u64,
    pub max_attempts: u32,
    pub initial_backoff_ms: u64,
    pub max_backoff_ms: u64,
    pub failure_threshold: u32,
    pub recovery_timeout_ms: u64,
    pub signature_header: String,
    pub event_header: String,
    pub delivery_id_header: String,
    pub timestamp_header: String,
}

impl Default for WebhookConfig {
    fn default() -> Self {
        Self {
            connect_timeout_ms: 5_000,
            read_timeout_ms: 30_000,
            max_attempts: 3,
            initial_backoff_ms: 1_000,
            max_backoff_ms: 60_000,
            failure_threshold: 5,
            recovery_timeout_ms: 300_000,
            signature_header: "X-MMF-Signature".into(),
            event_header: "X-MMF-Event".into(),
            delivery_id_header: "X-MMF-Delivery-Id".into(),
            timestamp_header: "X-MMF-Timestamp".into(),
        }
    }
}

impl WebhookConfig {
    pub fn validate(&self) -> Result<(), PushError> {
        if self.connect_timeout_ms == 0
            || self.read_timeout_ms == 0
            || self.max_attempts == 0
            || self.failure_threshold == 0
            || self.recovery_timeout_ms == 0
        {
            return Err(PushError::InvalidConfiguration(
                "webhook timeouts, attempts, and thresholds must be greater than zero".into(),
            ));
        }
        if [
            &self.signature_header,
            &self.event_header,
            &self.delivery_id_header,
            &self.timestamp_header,
        ]
        .iter()
        .any(|header| header.trim().is_empty())
        {
            return Err(PushError::InvalidConfiguration(
                "webhook header names must not be empty".into(),
            ));
        }
        Ok(())
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct WebhookRequest {
    pub url: String,
    pub headers: BTreeMap<String, String>,
    pub body: Vec<u8>,
    pub connect_timeout_ms: u64,
    pub read_timeout_ms: u64,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct WebhookResponse {
    pub status_code: u16,
    pub body: String,
    pub retry_after_seconds: Option<u64>,
}

#[async_trait]
pub trait WebhookProvider: Send + Sync {
    async fn start(&self) -> Result<(), PushError>;
    async fn stop(&self) -> Result<(), PushError>;
    async fn post(&self, request: &WebhookRequest) -> Result<WebhookResponse, PushError>;
    async fn health(&self) -> Result<(), PushError>;
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct WebhookCircuitStats {
    pub state: String,
}

pub struct WebhookAdapter {
    config: WebhookConfig,
    provider: Arc<dyn WebhookProvider>,
    configured_endpoints: Vec<WebhookEndpointConfig>,
    circuits: Mutex<BTreeMap<String, CircuitBreaker>>,
    backoff: ConfiguredBackoff,
    running: RwLock<bool>,
}

impl WebhookAdapter {
    pub fn new(
        config: WebhookConfig,
        provider: Arc<dyn WebhookProvider>,
        configured_endpoints: Vec<WebhookEndpointConfig>,
    ) -> Result<Self, PushError> {
        config.validate()?;
        for endpoint in &configured_endpoints {
            validate_endpoint(endpoint)?;
        }
        let backoff = ConfiguredBackoff::new(RetryConfig {
            max_attempts: config.max_attempts,
            base_delay_ms: config.initial_backoff_ms,
            max_delay_ms: config.max_backoff_ms,
            strategy: RetryStrategy::Exponential,
            backoff_multiplier: 2.0,
            jitter: false,
            jitter_factor: 0.0,
        })
        .map_err(|error| PushError::InvalidConfiguration(error.to_string()))?;
        Ok(Self {
            config,
            provider,
            configured_endpoints,
            circuits: Mutex::new(BTreeMap::new()),
            backoff,
            running: RwLock::new(false),
        })
    }

    #[must_use]
    pub fn sign_payload(body: &[u8], secret: &str) -> String {
        if secret.is_empty() {
            return String::new();
        }
        let Ok(mut mac) = HmacSha256::new_from_slice(secret.as_bytes()) else {
            return String::new();
        };
        mac.update(body);
        let digest = mac.finalize().into_bytes();
        let encoded = encode_hex(&digest);
        format!("sha256={encoded}")
    }

    #[must_use]
    pub fn verify_signature(body: &[u8], secret: &str, signature: &str) -> bool {
        let Some(encoded) = signature.strip_prefix("sha256=") else {
            return false;
        };
        let Some(bytes) = decode_hex(encoded) else {
            return false;
        };
        let Ok(mut mac) = HmacSha256::new_from_slice(secret.as_bytes()) else {
            return false;
        };
        mac.update(body);
        mac.verify_slice(&bytes).is_ok()
    }

    pub async fn send_to_endpoints(
        &self,
        message: &PushMessage,
        endpoints: &[WebhookEndpointConfig],
        now_ms: u64,
    ) -> Result<PushResult, PushError> {
        let event_type = message
            .data
            .get("event_type")
            .and_then(Value::as_str)
            .unwrap_or("");
        let filtered: Vec<_> = endpoints
            .iter()
            .filter(|endpoint| {
                endpoint.enabled
                    && (endpoint.event_types.is_empty()
                        || endpoint.event_types.iter().any(|item| item == event_type))
            })
            .collect();
        if filtered.is_empty() {
            let mut result = PushResult::delivered(&message.id, PushChannel::Webhook, now_ms);
            result.metadata.insert(
                "skipped".into(),
                Value::String("No matching endpoints for event type".into()),
            );
            return Ok(result);
        }
        let mut successes = 0_usize;
        let mut failures = Vec::new();
        for endpoint in &filtered {
            match self.deliver_to_endpoint(message, endpoint, now_ms).await {
                Ok(result) if result.success => successes = successes.saturating_add(1),
                Ok(result) => failures.push(result),
                Err(error) => failures.push(PushResult::failure(
                    &message.id,
                    PushChannel::Webhook,
                    PushStatus::Failed,
                    "PROVIDER_ERROR",
                    error.to_string(),
                    now_ms,
                )),
            }
        }
        let mut aggregate = if successes > 0 {
            PushResult::delivered(&message.id, PushChannel::Webhook, now_ms)
        } else {
            PushResult::failure(
                &message.id,
                PushChannel::Webhook,
                PushStatus::Failed,
                "ALL_ENDPOINTS_FAILED",
                "No webhook endpoint accepted the delivery",
                now_ms,
            )
        };
        aggregate
            .metadata
            .insert("total_endpoints".into(), Value::from(filtered.len()));
        aggregate
            .metadata
            .insert("success_count".into(), Value::from(successes));
        aggregate
            .metadata
            .insert("failure_count".into(), Value::from(failures.len()));
        Ok(aggregate)
    }

    #[must_use]
    pub fn circuit_stats(&self) -> BTreeMap<String, WebhookCircuitStats> {
        self.circuits
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner)
            .iter()
            .map(|(url, circuit)| {
                (
                    url.clone(),
                    WebhookCircuitStats {
                        state: match circuit.state() {
                            CircuitState::Closed => "closed",
                            CircuitState::Open => "open",
                            CircuitState::HalfOpen => "half_open",
                        }
                        .into(),
                    },
                )
            })
            .collect()
    }

    fn circuit(&self, url: &str) -> Result<CircuitBreaker, PushError> {
        let mut circuits = self
            .circuits
            .lock()
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        if let Some(circuit) = circuits.get(url) {
            return Ok(circuit.clone());
        }
        let circuit = CircuitBreaker::new(
            format!("webhook:{url}"),
            CircuitBreakerConfig {
                failure_threshold: self.config.failure_threshold,
                success_threshold: 1,
                open_timeout_ms: self.config.recovery_timeout_ms,
                ..CircuitBreakerConfig::default()
            },
        )
        .map_err(|error| PushError::InvalidConfiguration(error.to_string()))?;
        circuits.insert(url.into(), circuit.clone());
        Ok(circuit)
    }

    #[allow(clippy::too_many_lines)]
    pub async fn deliver_to_endpoint(
        &self,
        message: &PushMessage,
        endpoint: &WebhookEndpointConfig,
        now_ms: u64,
    ) -> Result<PushResult, PushError> {
        validate_endpoint(endpoint)?;
        let circuit = self.circuit(&endpoint.url)?;
        if circuit.allow_call_at(Instant::now()).is_err() {
            return Ok(PushResult::failure(
                &message.id,
                PushChannel::Webhook,
                PushStatus::Failed,
                "CIRCUIT_OPEN",
                "Circuit breaker is open for this endpoint",
                now_ms,
            ));
        }
        let body = serde_json::to_vec(message)
            .map_err(|error| PushError::Serialization(error.to_string()))?;
        let mut headers = BTreeMap::from([
            ("Content-Type".into(), "application/json".into()),
            (
                self.config.event_header.clone(),
                message
                    .data
                    .get("event_type")
                    .and_then(Value::as_str)
                    .unwrap_or("push")
                    .into(),
            ),
            (self.config.delivery_id_header.clone(), message.id.clone()),
            (self.config.timestamp_header.clone(), now_ms.to_string()),
        ]);
        headers.extend(endpoint.custom_headers.clone());
        let signature = Self::sign_payload(&body, &endpoint.secret);
        if !signature.is_empty() {
            headers.insert(self.config.signature_header.clone(), signature);
        }
        if let Some(correlation_id) = &message.correlation_id {
            headers.insert("X-MMF-Correlation-Id".into(), correlation_id.clone());
        }
        let request = WebhookRequest {
            url: endpoint.url.clone(),
            headers,
            body,
            connect_timeout_ms: self.config.connect_timeout_ms,
            read_timeout_ms: self.config.read_timeout_ms,
        };
        for attempt in 1..=self.config.max_attempts {
            match self.provider.post(&request).await {
                Ok(response) if response.status_code < 400 => {
                    circuit.record_success_at(Instant::now());
                    let mut result =
                        PushResult::delivered(&message.id, PushChannel::Webhook, now_ms);
                    result.attempt_number = attempt;
                    result
                        .metadata
                        .insert("status_code".into(), Value::from(response.status_code));
                    result
                        .metadata
                        .insert("endpoint".into(), Value::String(endpoint.url.clone()));
                    return Ok(result);
                }
                Ok(response) if response.status_code >= 500 => {
                    circuit.record_failure_at(Instant::now());
                    if attempt < self.config.max_attempts {
                        tokio::time::sleep(self.backoff.delay(attempt)).await;
                        continue;
                    }
                    let mut result = PushResult::failure(
                        &message.id,
                        PushChannel::Webhook,
                        PushStatus::Failed,
                        response.status_code.to_string(),
                        response.body,
                        now_ms,
                    );
                    result.attempt_number = attempt;
                    result.should_retry = true;
                    result.retry_after_seconds = response.retry_after_seconds;
                    return Ok(result);
                }
                Ok(response) => {
                    let mut result = PushResult::failure(
                        &message.id,
                        PushChannel::Webhook,
                        PushStatus::Rejected,
                        response.status_code.to_string(),
                        response.body.chars().take(200).collect::<String>(),
                        now_ms,
                    );
                    result.attempt_number = attempt;
                    return Ok(result);
                }
                Err(error) => {
                    circuit.record_failure_at(Instant::now());
                    if attempt < self.config.max_attempts {
                        tokio::time::sleep(self.backoff.delay(attempt)).await;
                        continue;
                    }
                    let mut result = PushResult::failure(
                        &message.id,
                        PushChannel::Webhook,
                        PushStatus::Failed,
                        "PROVIDER_ERROR",
                        error.to_string(),
                        now_ms,
                    );
                    result.attempt_number = attempt;
                    result.should_retry = true;
                    return Ok(result);
                }
            }
        }
        Err(PushError::Delivery("webhook attempts exhausted".into()))
    }
}

#[async_trait]
impl PushAdapter for WebhookAdapter {
    fn channel(&self) -> PushChannel {
        PushChannel::Webhook
    }

    async fn start(&self) -> Result<(), PushError> {
        self.provider.start().await?;
        *self
            .running
            .write()
            .unwrap_or_else(std::sync::PoisonError::into_inner) = true;
        Ok(())
    }

    async fn stop(&self) -> Result<(), PushError> {
        self.provider.stop().await?;
        *self
            .running
            .write()
            .unwrap_or_else(std::sync::PoisonError::into_inner) = false;
        Ok(())
    }

    async fn send(&self, message: &PushMessage, now_ms: u64) -> Result<PushResult, PushError> {
        let endpoints = if message.target.webhook_urls.is_empty() {
            self.configured_endpoints.clone()
        } else {
            message
                .target
                .webhook_urls
                .iter()
                .map(|url| WebhookEndpointConfig {
                    url: url.clone(),
                    enabled: true,
                    ..WebhookEndpointConfig::default()
                })
                .collect()
        };
        if endpoints.is_empty() {
            return Ok(PushResult::failure(
                &message.id,
                PushChannel::Webhook,
                PushStatus::Failed,
                "NO_ENDPOINTS",
                "No webhook URLs provided",
                now_ms,
            ));
        }
        self.send_to_endpoints(message, &endpoints, now_ms).await
    }

    async fn health(&self, now_ms: u64) -> PushAdapterHealth {
        let running = *self
            .running
            .read()
            .unwrap_or_else(std::sync::PoisonError::into_inner);
        let (status, detail) = if running {
            match self.provider.health().await {
                Ok(()) => (PushHealthStatus::Healthy, None),
                Err(error) => (PushHealthStatus::Unavailable, Some(error.to_string())),
            }
        } else {
            (PushHealthStatus::Stopped, None)
        };
        PushAdapterHealth {
            channel: PushChannel::Webhook,
            status,
            detail,
            checked_at_ms: now_ms,
        }
    }
}

fn validate_endpoint(endpoint: &WebhookEndpointConfig) -> Result<(), PushError> {
    validate_destination_shape(&endpoint.url)
}

fn validate_destination_shape(value: &str) -> Result<(), PushError> {
    let parse_value = value.replace(TOKEN_MARKER, "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
    let parsed = url::Url::parse(&parse_value).map_err(|_| {
        PushError::InvalidConfiguration("webhook endpoint must be an absolute URL".into())
    })?;
    if !matches!(parsed.scheme(), "http" | "https")
        || parsed.host_str().is_none()
        || !parsed.username().is_empty()
        || parsed.password().is_some()
        || parsed.fragment().is_some()
    {
        return Err(PushError::InvalidConfiguration(
            "webhook endpoint must be HTTP(S) without userinfo or fragments".into(),
        ));
    }
    Ok(())
}

fn destination_matches(destination: &str, template: &str) -> bool {
    let Some((prefix, suffix)) = template.split_once(TOKEN_MARKER) else {
        return destination == template;
    };
    let Some(token_with_suffix) = destination.strip_prefix(prefix) else {
        return false;
    };
    let Some(token) = token_with_suffix.strip_suffix(suffix) else {
        return false;
    };
    (16..=512).contains(&token.len())
        && token
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'-'))
}

fn decode_hex(value: &str) -> Option<Vec<u8>> {
    if !value.len().is_multiple_of(2) {
        return None;
    }
    (0..value.len())
        .step_by(2)
        .map(|index| u8::from_str_radix(&value[index..index + 2], 16).ok())
        .collect()
}

const fn default_true() -> bool {
    true
}
